rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+use crate::ffi;
+use std::borrow::Borrow;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::CStr;
+use std::ffi::OsStr;
+#[cfg(windows)]
+use std::mem::forget;
+use std::mem::ManuallyDrop;
+use std::ptr::null;
+use std::ptr::NonNull;
+use std::str::Utf8Error;
+
+// This is more or less equivalent to manually defining Display and From<other error types>
+use thiserror::Error;
```

Comment:
I used thiserror to make the original sketch easier, do you think it's a reasonable dependency for something like this to pull in? Otherwise we could write it out ourselves with a little bit of boilerplate.

Reply:
I think having thiserror is reasonable. Especially since it is not exposed in any way.